sdk/protocol2.go

```diff
@@ -0,0 +1,173 @@
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/newrelic/infra-integrations-sdk/args"
+	"github.com/newrelic/infra-integrations-sdk/cache"
+	"github.com/newrelic/infra-integrations-sdk/log"
+	"github.com/newrelic/infra-integrations-sdk/metric"
+)
+
+// Entity is the producer of the data. Entity could be a host, a container, a pod, or whatever unit of meaning.
+type Entity struct {
+	Name string `json:"name"`
+	Type string `json:"type"`
+}
+
+// EntityData defines all the data related to a particular event from an entity.
+type EntityData struct {
+	Entity    Entity             `json:"entity"`
+	Metrics   []metric.MetricSet `json:"metrics"`
+	Inventory Inventory          `json:"inventory"`
+	Events    []Event            `json:"events"`
+}
+
+// EntitiesData is a collection of EntityData.
+type EntitiesData []*EntityData
+
+// Has tells you if that Entity is already part of the data.
+func (d EntitiesData) Has(entity Entity) bool {
+	for _, e := range d {
+		if e.Entity.Name == entity.Name && e.Entity.Type == entity.Type {
+			return true
+		}
+	}
+
+	return false
+}
+
+// IntegrationProtocol2 defines the format of the output JSON that integrations will return for protocol 2.
+type IntegrationProtocol2 struct {
+	sync.Mutex
+	Name               string       `json:"name"`
+	ProtocolVersion    string       `json:"protocol_version"`
+	IntegrationVersion string       `json:"integration_version"`
+	Data               EntitiesData `json:"data"`
+	prettyOutput       bool
+	writer             io.Writer
+}
+
+// NewIntegrationProtocol2WithWriter initializes a new instance of the integration protocol 2 specifying a writer instead of using stdout by default.
+func NewIntegrationProtocol2WithWriter(name string, version string, arguments interface{}, writer io.Writer) (*IntegrationProtocol2, error) {
+	i, err := NewIntegrationProtocol2(name, version, arguments)
+	if err != nil {
+		return nil, err
+	}
+
+	i.writer = writer
+
+	return i, nil
+}
+
+// NewIntegrationProtocol2 initializes a new instance of the integration protocol 2.
+func NewIntegrationProtocol2(name string, version string, arguments interface{}) (*IntegrationProtocol2, error) {
+	err := args.SetupArgs(arguments)
+	if err != nil {
+		return nil, err
+	}
+	defaultArgs := args.GetDefaultArgs(arguments)
+
+	log.SetupLogging(defaultArgs.Verbose)
+
+	// Avoid working with an uninitialized or in error state cache
+	if err = cache.Status(); err != nil {
+		return nil, err
+	}
+
+	integration := &IntegrationProtocol2{
+		Name:               name,
+		ProtocolVersion:    "2",
+		IntegrationVersion: version,
+		Data:               make([]*EntityData, 0), // TODO should we make?
+		prettyOutput:       defaultArgs.Pretty,
+		writer:             os.Stdout, // StdOut by default
+	}
+
+	return integration, nil
+}
+
+// Entity method creates a new EntityData and returns it if it does not exists.
+func (integration *IntegrationProtocol2) Entity(entityName, entityType string) (*EntityData, error) {
```

Comment:
I would call it `NewEntity`.

Should we have the functionality for updating `Entity`?

Reply:
The idea of this method is create or retrieve an already existing Entity. Since it's not possible to do it now (internal reasons), we could do it in the future without introducing any breaking change in the name.